crates/bevy_ecs/src/schedule/set.rs

```diff
@@ -5,61 +5,147 @@ use std::marker::PhantomData;
 use std::sync::atomic::{AtomicUsize, Ordering};
 
 pub use bevy_ecs_macros::{ScheduleLabel, SystemSet};
-use bevy_utils::define_boxed_label;
 use bevy_utils::label::DynHash;
 
 use crate::system::{
     ExclusiveSystemParamFunction, IsExclusiveFunctionSystem, IsFunctionSystem, SystemParamFunction,
 };
 
-define_boxed_label!(ScheduleLabel);
-
-/// A shorthand for `Box<dyn SystemSet>`.
-pub type BoxedSystemSet = Box<dyn SystemSet>;
-/// A shorthand for `Box<dyn ScheduleLabel>`.
-pub type BoxedScheduleLabel = Box<dyn ScheduleLabel>;
-
-/// Types that identify logical groups of systems.
+/// Identifies a logical group of systems.
 pub trait SystemSet: DynHash + Debug + Send + Sync + 'static {
-    /// Returns `Some` if this system set is a [`SystemTypeSet`].
+    /// Returns the [`TypeId`] of the system if the system set is a [`SystemTypeSet`].
+    ///
+    /// A [`SystemTypeSet`] has special properties:
+    /// - You cannot manually add systems or sets to it.
+    /// - You cannot configure it.
+    /// - You cannot order relative to it if it contains more than one instance.
+    ///
+    /// These sets are automatically populated, so these constraints exist to prevent unintentional ambiguity.
     fn system_type(&self) -> Option<TypeId> {
         None
     }
 
-    /// Returns `true` if this system set is an [`AnonymousSet`].
+    /// Returns `true` if the system set is an [`AnonymousSet`].
     fn is_anonymous(&self) -> bool {
         false
     }
-    /// Creates a boxed clone of the label corresponding to this system set.
-    fn dyn_clone(&self) -> Box<dyn SystemSet>;
+
+    /// Returns the unique, type-elided identifier for the system set.
+    fn as_label(&self) -> SystemSetId {
+        SystemSetId::of(self)
+    }
+
+    /// Returns the type-elided version of the system set.
+    fn as_untyped(&self) -> SystemSetUntyped {
+        SystemSetUntyped::of(self)
+    }
 }
 
-impl PartialEq for dyn SystemSet {
-    fn eq(&self, other: &Self) -> bool {
-        self.dyn_eq(other.as_dyn_eq())
+/// A lightweight and printable identifier for a [`SystemSet`].
+#[derive(Clone, Copy, Eq)]
+pub struct SystemSetId(&'static str);
```

Comment:
How come you don't use the `define_label!` macro for `SystemSet`?

Reply:
`SystemSet` has more methods than other labels. There's a distinction between normal system sets, system type sets, and anonymous system sets.